src/proto/streams/state.rs

```diff
@@ -111,7 +111,7 @@ impl State {
                     Open { local, remote }
                 }
             }
-            HalfClosedRemote(AwaitingHeaders) | ReservedLocal => {
+            HalfClosedRemote(AwaitingHeaders | Streaming) | ReservedLocal => {
```

Comment:
I spent some time thinking about this in particular. There is no documentation on the struct hinting at how to reason about the state, but if I understood it correctly `HalfClosedRemote` means that we've gotten an EOS from the remote side. `HalfClosedRemote(AwaitingHeaders)` this side currently haven't written headers yet (the enum-name `Peer` is a bit misleading here?). This used to encode the rule that headers should only be sent _once_. With 100-continue in the picture, however, that doesn't really hold true. The easiest way I could see to model this, is to simply relax this rule as proposed here. Perhaps another approach could be feasible? I.E. something not "Opening the send-half", but instead explicitly "finish sending headers"?

Reply:
The code comments at the top of file include the flow chart from the spec, but the the spec includes more description of each possible state: https://httpwg.org/specs/rfc7540.html#StreamStates.

It looks like your understanding is correct. I've also found the usage of `Peer` in this crate to be the opposite of what I usually think, but it is consistently used in this file and elsewhere to mean "me" (I usually think of peer as the remote, but I didn't come up with those names :shrug:).

Reply:
However, I think we don't want to change the state here, since sending a `HEADERS` frame with a 1xx code shouldn't mean it is now waiting to send `DATA` frames. It still needs to send the "final" code in another `HEADERS` frame.